contracts/asset-proxy/contracts/src/bridges/UniswapBridge.sol

```diff
@@ -0,0 +1,229 @@
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-erc20/contracts/src/interfaces/IERC20Token.sol";
+import "@0x/contracts-erc20/contracts/src/interfaces/IEtherToken.sol";
+import "@0x/contracts-exchange-libs/contracts/src/IWallet.sol";
+import "../interfaces/IUniswapExchangeFactory.sol";
+import "../interfaces/IUniswapExchange.sol";
+import "../interfaces/IERC20Bridge.sol";
+
+
+// solhint-disable space-after-comma
+// solhint-disable not-rely-on-time
+contract UniswapBridge is
+    IERC20Bridge,
+    IWallet
+{
+    /* Mainnet addresses */
+    address constant private UNISWAP_EXCHANGE_FACTORY_ADDRESS = 0xc0a47dFe034B400B47bDaD5FecDa2621de6c4d95;
+    address constant private WETH_ADDRESS = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    // Struct to hold `withdrawTo()` local variables in memory and to avoid
+    // stack overflows.
+    struct WithdrawToState {
+        address exchangeTokenAddress;
+        IUniswapExchange exchange;
+        uint256 fromTokenBalance;
+        IEtherToken weth;
+    }
+
+    // solhint-disable no-empty-blocks
+    /// @dev Payable fallback to receive ETH from uniswap.
+    function ()
+        external
+        payable
+    {}
+
+    /// @dev Callback for `IERC20Bridge`. Tries to buy `amount` of
+    ///      `toTokenAddress` tokens by selling the entirety of the `fromTokenAddress`
+    ///      token encoded in the bridge data.
+    /// @param toTokenAddress The token to buy and transfer to `to`.
+    /// @param to The recipient of the bought tokens.
+    /// @param amount Minimum amount of `toTokenAddress` tokens to buy.
+    /// @param bridgeData The abi-encoded "from" token address.
+    /// @return success The magic bytes if successful.
+    function withdrawTo(
+        address toTokenAddress,
+        address /* from */,
+        address to,
+        uint256 amount,
+        bytes calldata bridgeData
+    )
+        external
+        returns (bytes4 success)
+    {
+        // State memory object to avoid stack overflows.
+        WithdrawToState memory state;
+        // Decode the bridge data to get the `fromTokenAddress`.
+        (address fromTokenAddress) = abi.decode(bridgeData, (address));
+
+        // Just transfer the tokens if they're the same.
+        if (fromTokenAddress == toTokenAddress) {
+            IERC20Token(fromTokenAddress).transfer(to, amount);
+            return BRIDGE_SUCCESS;
+        }
+
+        // Get the  exchange token for the token pair.
+        state.exchangeTokenAddress = _getUniswapExchangeTokenAddressForTokenPair(
+            fromTokenAddress,
+            toTokenAddress
+        );
+        // Get the exchange.
+        state.exchange = _getUniswapExchangeForToken(state.exchangeTokenAddress);
+        // Grant an allowance to the exchange.
+        _grantExchangeAllowance(state.exchange, state.exchangeTokenAddress);
```

Comment:
Shouldn't this be `_grantExchangeAllowance(state.exchange, fromTokenAddress)`? We also shouldn't need to call this if the `fromTokenAddress` is weth.

Reply:
Ah, yeah, that's true. `state.exchangeTokenAddress` will always be `fromTokenAddress`. :+1: